src/pixelator/pixeldataset/precomputed_layouts.py

```diff
@@ -33,8 +33,169 @@ class PreComputedLayoutsEmpty(PixelatorBaseException):
     """Raised when trying to access an empty PreComputedLayouts instance."""
 
 
+class RequestedLayoutsHaveMultipleDimensionalities(PixelatorBaseException):
+    """Raised when trying to materialize PreComputedLayouts with multiple dimensionalities."""
+
+
+class _DataProvider(Protocol):
+    def is_empty(self) -> bool: ...
+
+    def to_df(self, columns: list[str] | None = None) -> pd.DataFrame: ...
+
+    def lazy(self): ...
+
+    def unique_components(self): ...
+
+    def filter(
+        self,
+        component_ids: str | set[str] | None = None,
+        graph_projections: str | set[str] | None = None,
+        layout_methods: str | set[str] | None = None,
+    ) -> pl.LazyFrame | list[pl.LazyFrame]: ...
+
+
+class _EmptyDataProvider(_DataProvider):
+    def __init__(self) -> None:
+        # This class needs no parameters
+        pass
+
+    def is_empty(self) -> bool:
+        return True
+
+    def to_df(self, columns: list[str] | None = None) -> pd.DataFrame:
+        raise PreComputedLayoutsEmpty("No layouts available")
+
+    def lazy(self):
+        raise PreComputedLayoutsEmpty("No layouts available")
+
+    def unique_components(self):
+        return {}
+
+    def filter(
+        self,
+        component_ids: str | set[str] | None = None,
+        graph_projections: str | set[str] | None = None,
+        layout_methods: str | set[str] | None = None,
+    ) -> list[pl.LazyFrame]:
+        raise PreComputedLayoutsEmpty("No layouts available")
+
+
+class _SingleFrameDataProvider(_DataProvider):
+    def __init__(self, lazy_frame: pl.LazyFrame) -> None:
+        self._lazy_frame = lazy_frame
+
+    def is_empty(self) -> bool:
+        return self.lazy().select(pl.col("component")).first().collect().is_empty()
```

Comment:
As we are using in all the other ones?
```suggestion
        return self._lazy_frame.select(pl.col("component")).first().collect().is_empty()
```

Reply:
Good catch.